CPS-0003/README.md

```diff
@@ -0,0 +1,141 @@
+
+UTxOs have basically 2 operations associated with them:
+
+- READ - by reference inputs
+- SPEND - whenever the utxo is an input
+
+From these only the SPEND is controlled by smart-contracts. Currently reference inputs cannot control the transactions they are in (which is a topic for another CPS).
+
+In comparison, the operations with tokens are 3 fold:
+
+- mint
+- burn
+- transfer
+
+The MINT plutus scripts control mint and burn, but not transfer. There are 2 possible directions how the transfer could be validated:
+
+1. Call the token policy with the `MINT`, also whenever a smart token is transferred. The minting policy then could decide based on the redeemer for how it should validate the transaction.
+2. Introduce a `TRANSFER` operation and a separate script type that would be used for this exact use-case.
+
+Both of these options would strain wallets and dapps, as they would need additional specific logic to handle smart tokens. The token validation scripts might not be generic enough and might be costly to do in every interaction. As the blockchain evolves, some interactions couldn't be predicted by the token and evolving the scripts would be problematic.
+
+A workaround might be, if some smart-token policies delegated the validation to some validations scripts. Unfortunately, without adding the ability to execute validation on referenced scripts, even this would require either: minting/forging a token in the transaction or spending an utxo with the desired validation logic. This approach doesn't seem scalable-enough, either.
+
+### Identifying smart tokens
+
+Drawing parallels with addresses, on cardano addresses have a 1 byte header identifying if they are a script address, pubkey address, reward address, etc. Unfortunately this is not the case with tokens currently. Policy IDs unfortunately do not have headers and it's assumed that it's always the hash of a native or plutus script.
+
+The identification is necessary, so that the nodes knows when it would require an additional transfer script to be present in the transaction. There might be 3 directions how "smart" or "validated" tokens could be identified in transactions:
+
+**1. segregate tokens**
+In this case, the validated and smart tokens could be separated in the transaction structures. In the CDDL:
+
+```cddl
+value = coin / [coin,multiasset<uint>] / [coin, multiasset<uint>, smartasset<uint>]
+mint = multiasset<int64> / [multiasset<uint>, smartasset<uint>]
+```
+
+Alternatively the value could be "objectified" similar to babbage tx outputs:
+
+```cddl
+value = {
+  0: coin,
+  ? 1: multiasset<uint>,
+  ? 2: smartasset<uint>
+}
+```
+
+This explicit segregation would have the benefit that it could be easier backwards compatible. In this format also the policy ids could remain as `hash28` script hashes.
+
+**2. add policy headers**
+This would be a breaking change to the structures and policies. Basically it would be redefining the multiasset structure to:
+
+```
```

Comment:
```suggestion
```cddl
```

Reply:
great; I don't think we had the `cddl` GitHub language type 2 years ago 😅